src/error.rs

```diff
@@ -340,6 +343,7 @@ pub struct WriteConcernError {
 /// An error that occurred during a write operation that wasn't due to being unable to satisfy a
 /// write concern.
 #[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
 pub struct WriteError {
```

Comment:
Is there a macro for disallowing initialization? If not, we may want to consider exposing these fields via methods instead, since any new fields we add to this would need to have a default implementation to avoid breaking user code, which for something like an error struct may not make sense.

Reply:
Normally the solution is to just have private fields and define getters. If we're feeling especially lazy, we could use the [`getset` crate](https://crates.io/crates/getset).

Reply:
So rereading the `non_exhaustive` documentation, this macro actually makes it so outside crates cannot instantiate these types--they can only match on them. This works great for types like this, but not so great for our options types. We may have to only allow users to use the builder for creating options then. We also need to revisit the changes to BSON to make sure outside users can still create the things they need to.

Reply:
Interesting; I'm not sure I like the idea of requiring the builder pattern for all options structs, as some of them are pretty small, making it overkill. I'll remove it from types we want users to be able to create for now until we reach a decision.

Reply:
I left it on all enums (as a top-level attribute, not on the variants) since they can still be constructed even with it, as well as on all error- and event-related structs, since they only need to be emitted by the driver rather than created by users.

Reply:
So we actually don't need the getters and setters on the various error cases anymore, since `non_exhaustive` gives us the non-construction. I apologize for realizing this after you implemented all of them...

As for options, I'm not sure there's any other option (pun intended). If we allow users do do: 
```rust
let options = FooOptions { a: "hello".to_string() };
```
then any new field added to `FooOptions` will be a breaking change, which definitely won't work for us. We could add other methods to `FooOptions` that are constructor like, but I think the builder pattern is probably the best way to do that.

Reply:
Getters removed and `non_exhaustive` added to all options structs as discussed